Description: Ensure reproducible builds by evaluating directory paths at configure time
 OPENSSLDIR, ENGINESDIR and MODULESDIR are passed to the compiler as -D flags
 referencing Make variables (e.g. $(OPENSSLDIR)). These Make variables can
 potentially be overridden at build time, breaking reproducibility. Fix this by
 resolving the paths at configure time using the Perl variables that are already
 available, embedding fixed literal strings in the generated Makefile.
Author: Pexip AS <packaging@pexip.com>
Forwarded: not-needed
Last-Update: 2026-08-06

--- a/Configurations/unix-Makefile.tmpl
+++ b/Configurations/unix-Makefile.tmpl
@@ -449,10 +449,15 @@
                           @{$config{lib_cppflags}},
                           @{$config{shared_cppflag}});
                 join(' ', $lib_cppflags,
-                          (map { '-D'.$_ }
-                               'OPENSSLDIR="\"$(OPENSSLDIR)\""',
-                               'ENGINESDIR="\"$(ENGINESDIR)\""',
-                               'MODULESDIR="\"$(MODULESDIR)\""'),
+                          (map { '-D'.$_ } do {
+                               use File::Spec::Functions;
+                               my $resolved_libdir = file_name_is_absolute($libdir)
+                                   ? $libdir : catdir($prefix, $libdir);
+                               my $engdir = $config{enginesdir} || catdir($resolved_libdir, "engines-$sover_dirname");
+                               my $moddir = $config{modulesdir} || catdir($resolved_libdir, "ossl-modules");
+                               ("OPENSSLDIR=\"\\\"$openssldir\\\"\"",
+                                "ENGINESDIR=\"\\\"$engdir\\\"\"",
+                                "MODULESDIR=\"\\\"$moddir\\\"\"") }),
                           '$(CNF_CPPFLAGS)', '$(CPPFLAGS)') -}
 LIB_CFLAGS={- join(' ', $target{lib_cflags} || (),
                         $target{shared_cflag} || (),
--- a/util/mkbuildinf.pl
+++ b/util/mkbuildinf.pl
@@ -13,6 +13,10 @@
 my $platform = pop @ARGV;
 my $cflags = join(' ', @ARGV);
 $cflags =~ s(\\)(\\\\)g;
+
+# Strip include paths and directory defines that may contain build-tree paths
+$cflags =~ s/\s*-I\S+//g;
+
 $cflags = "compiler: $cflags";
 
 # Use the value of the envvar SOURCE_DATE_EPOCH, even if it's
